File: 2D-World-Simulator-Java/OPP-Java/src/op/JavaProject/Game/Organism.java
package op.JavaProject.Game;

import op.JavaProject.Functionality.BoardPoint;
import op.JavaProject.Functionality.Species;
import op.JavaProject.Organisms.Animals.*;
import op.JavaProject.Organisms.Plants.*;

import java.awt.*;
import java.io.Serializable;
import java.util.Random;

import static op.JavaProject.Game.OrganismsStatics.*;

public abstract class Organism implements Serializable {

    protected int strength;

    protected int initiative;
    protected int whenBorn;
    protected boolean killed;

    protected World organismWorld;
    protected BoardPoint organismPos;

    protected Species organismSpecie;

    protected String organismSign;

    protected Color organismColor;

    protected Organism(World organismWorld, BoardPoint point) {
        this.organismWorld = organismWorld;
        this.organismPos = new BoardPoint(point.GetX(), point.GetY());
        this.whenBorn = organismWorld.GetTurnNumber();
    }

    //moving organism to given position
    protected void MoveToGivenField(BoardPoint newPoint) {
        organismWorld.GetStatesOfFields()[organismPos.GetY()][organismPos.GetX()] = null;
        organismWorld.GetStatesOfFields()[newPoint.GetY()][newPoint.GetX()] = this;
        organismPos.SetX(newPoint.GetX());
        organismPos.SetY(newPoint.GetY());
    }

    //getting random field near organism, it may be free or not
    //point may stay the same -> f.e. if there is no place for spawning new organism
    public BoardPoint GetNearPoint(BoardPoint point, boolean free) {

        BoardPoint newPoint = new BoardPoint();
        int move, posX = point.GetX(), posY = point.GetY(), newPosX = posX, newPosY = posY;
        boolean correctPos = false;
        boolean leftCheck = false, rightCheck = false, upCheck = false, downCheck = false, upSecond = false, downSecond = false;
        int sides = SIDES, i = 0, checked = 0;
        boolean hex = organismWorld.GetIfHex();
        ;
        if (hex) sides = SIDES_HEX;
        do {
            if (((checked > (SIDES - 1)) && !hex) || ((checked > (SIDES_HEX - 1)) && hex)) break;
            Random randomValue = new Random();

            if (hex) {
                move = randomValue.nextInt(SIDES_HEX);
            } else {
                move = randomValue.nextInt(SIDES);
            }

            i = 0;
            newPosX = posX;
            newPosY = posY;

            if (!upCheck) {
                if (move == UP) {
                    newPosY = posY + 1;
                    upCheck = true;
                }
                i++;
            }

            if (!downCheck) {
                if (move == DOWN) {
                    newPosY = posY - 1;
                    downCheck = true;
                }
                i++;
            }

            if (!rightCheck) {
                if (move == RIGHT) {
                    newPosX = posX + 1;
                    rightCheck = true;
                }
                i++;
            }

            if (!leftCheck) {
                if (move == LEFT) {
                    newPosX = posX - 1;
                    leftCheck = true;
                }
                i++;
            }

            //if it is hexagonal world we also need to check additional fields
            if (hex) {
                if (!downSecond) {

                    if (move == SECOND_DOWN) {
                        if (newPosY % 2 == 0) {
                            newPosX = posX - 1;
                            newPosY = posY + 1;
                        } else {
                            newPosX = posX + 1;
                            newPosY = posY + 1;
                        }

                        downSecond = true;
                    }
                    i++;
                }

                if (!upSecond) {
                    if (move == SECOND_UP) {
                        if (newPosY % 2 == 0) {
                            newPosX = posX - 1;
                            newPosY = posY - 1;
                        } else {
                            newPosX = posX + 1;
                            newPosY = posY - 1;
                        }

                        upSecond = true;
                    }
                    i++;
                }
            }

            sides--;

            if (CheckPointCorrectness(newPosX, newPosY)) {
                if (free) {
                    //that means we checked all sides and there is no possible free side
                    if (leftCheck && rightCheck && upCheck && downCheck) {
                        if (hex && downSecond && upSecond) {
                            //no place around, point stays the same
                            break;
                        }
                        //point stays the same
                        if (!hex) break;
                    } else {
                        //we also need to check if the field is free (the point)
                        if (CheckIfFieldIsFree(newPosX, newPosY)) {
                            correctPos = true;
                        }

                    }
                } else {
                    correctPos = true;
                }
            }

        } while (!correctPos);

        if (correctPos) {
            newPoint.SetX(newPosX);
            newPoint.SetY(newPosY);
            return newPoint;
        } else {
            //no place around, point stays the same
            return organismPos;
        }

    }

    //checking if point is on the map
    public boolean CheckPointCorrectness(int x, int y) {
        if (y < organismWorld.GetHeight() && y >= 0) {
            if (x < organismWorld.GetWidth() && x >= 0) {
                return true;
            }
        }
        return false;
    }

    abstract public void Action();

    abstract public void Collision(Organism otherOrganism);

    abstract public boolean Defence(Organism attackingOrganism);

    abstract public boolean CheckIfOrganismIsAnimal();

    //checking if given field is not occupied by other organism
    protected boolean CheckIfFieldIsFree(int x, int y) {

        if (organismWorld.GetStatesOfFields()[y][x] != null) {
            return false;
        }
        return true;

    }

    public void SetOrganismWorld(World organismWorld) {
        this.organismWorld = organismWorld;
    }

    //printing organism on the grid
    public void Draw() {
        if (!organismWorld.GetIfHex()) {
            organismWorld.GetWorldFrame().boardPanel.board[organismPos.GetY()][organismPos.GetX()].setText(organismSign);
            organismWorld.GetWorldFrame().boardPanel.board[organismPos.GetY()][organismPos.GetX()].setEnabled(false);
            organismWorld.GetWorldFrame().boardPanel.board[organismPos.GetY()][organismPos.GetX()].setForeground(Color.BLACK);
            organismWorld.GetWorldFrame().boardPanel.board[organismPos.GetY()][organismPos.GetX()].setBackground(organismColor);
        } else {
            organismWorld.GetWorldFrame().boardPanelHex.hexButtons[organismPos.GetY()][organismPos.GetX()].changeColor(organismColor);
            organismWorld.GetWorldFrame().boardPanelHex.hexButtons[organismPos.GetY()][organismPos.GetX()].repaint();
            organismWorld.GetWorldFrame().boardPanelHex.hexButtons[organismPos.GetY()][organismPos.GetX()].setEnabled(false);
        }

    }

    public BoardPoint GetOrganismPos() {
        return organismPos;
    }

    //starting game function, creating every kind of organism once
    public static void CreateEveryOrganism(World organismWorld) {
        //Creating animals
        BoardPoint newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Wolf(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Sheep(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Fox(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Turtle(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Antelope(organismWorld, newPoint));

        //Creating plants
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Grass(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new SowThistle(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Guarana(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new Belladonna(organismWorld, newPoint));
        newPoint = organismWorld.RandomFreePoint();
        organismWorld.AddOrganismToWorld(new SosnowskysHogweed(organismWorld, newPoint));
    }

    //creating new organism depending on given specie and position
    public static Organism CreateNewOrganism(Species organismSpecie, World organismWorld, BoardPoint organismPoint) {
        switch (organismSpecie) {
            case WOLF:
                return new Wolf(organismWorld, organismPoint);
            case SHEEP:
                return new Sheep(organismWorld, organismPoint);
            case FOX:
                return new Fox(organismWorld, organismPoint);
            case HUMAN:
                return new Human(organismWorld, organismPoint);
            case ANTELOPE:
                return new Antelope(organismWorld, organismPoint);
            case GRASS:
                return new Grass(organismWorld, organismPoint);
            case SOW_THISTLE:
                return new SowThistle(organismWorld, organismPoint);
            case GUARANA:
                return new Guarana(organismWorld, organismPoint);
            case BELLADONNA:
                return new Belladonna(organismWorld, organismPoint);
            case SOSNOWSKYS_HOGWEED:
                return new SosnowskysHogweed(organismWorld, organismPoint);

            default:
                return null;
        }
    }

    public Species GetSpecie() {
        return organismSpecie;
    }

    public boolean GetKilled() {
        return killed;
    }

    public int GetInitiative() {
        return initiative;
    }

    public int GetWhenBorn() {
        return whenBorn;
    }

    public String GetOrganismSign() {
        return organismSign;
    }

    public int GetStrength() {
        return strength;
    }

    public void SetStrength(int strength) {
        this.strength = strength;
    }

    public void SetKilled(boolean killed) {
        this.killed = killed;
    }
}

File: 2D-World-Simulator-CPP/OPP-CPP/Organism.cpp
#include "Organism.h"
#include "Wolf.h"
#include "Sheep.h"
#include "Human.h"
#include "Fox.h"
#include "Turtle.h"
#include "Antelope.h"
#include "Grass.h"
#include "SowThistle.h"
#include "Guarana.h"
#include "BellaDonna.h"
#include "SosnowskysHogweed.h"

Organism::Organism(World* world, Point point) {
	this->organismWorld = world;
	this->organismPos = new Point(point.getX(), point.getY());
	this->whenBorn = world->getTurnNumber();
}

OrganismSpecie Organism::getSpecie() const {
	return specie;
}

int Organism::getWhenBorn() const {
	return whenBorn;
}

Point Organism::getOrganismPos() const {
	return *organismPos;
}

int Organism::getInitiative() const{ 
	return initiative; 
}

int Organism::getStrength() const{
	return strength;
}

char Organism::getOrganismSign() const {
	return organismSign;
}

bool Organism::getKilled() const {
	return killed;
}

void Organism::setWhenBorn(int whenBorn) {
	this->whenBorn = whenBorn;
}

void Organism::setInitiative(int initiative) {
	this->initiative = initiative;
}

void Organism::setStrength(int strength) {
	this->strength = strength;
}

void Organism::setOrganismSpecie(OrganismSpecie specie) {
	this->specie = specie;
}

void Organism::setOrganismSign(char organismSign) {
	this->organismSign = organismSign;
}

void Organism::setKilled(bool killed) {
	this->killed = killed;
}

//moving organism to given position
void Organism::moveToGivenField(Point newPoint) {
	organismWorld->getStatesOfFields()[organismPos->getY()][organismPos->getX()] = nullptr;
	organismWorld->getStatesOfFields()[newPoint.getY()][newPoint.getX()] = this;
	organismPos->setX(newPoint.getX());
	organismPos->setY(newPoint.getY());
}

//getting random field near organism, it may be free or not
//point may stay the same -> f.e. if there is no place for resping new organism
Point Organism::getNearPoint(Point point, bool free) {

	Point newPoint;
	int move, posX = point.getX(), posY = point.getY(), newPosX = posX, newPosY = posY;
	bool correctPos = false;
	bool leftCheck = false, rightCheck = false, upCheck = false, downCheck = false;
	int sides = SIDES, i = NULL;
	do {
		if (sides == NULL) break;
		move = rand() % sides;
		i = NULL;
		newPosX = posX;
		newPosY = posY;

		if (!rightCheck) {
			if (move == i) newPosX = posX + ONE, rightCheck = true;
			i++;
		}
		
		if (!leftCheck) {
			if (move == i) newPosX = posX - ONE, leftCheck = true;
			i++;
		}
		
		if (!upCheck) {
			if (move == i) newPosY = posY + ONE, upCheck = true;
			i++;
		}
		
		if (!downCheck) {
			if (move == i) newPosY = posY - ONE, downCheck = true;
			i++;
		}
		
		sides--;

		if (checkPointCorrenctness(newPosX, newPosY)) {
			if (free) {
				//that means we checked all sides and there is no possible free side
				if (leftCheck && rightCheck && upCheck && downCheck) {
					//no place around, point stays the same
					break;
				}
				else {
					//we also need to check if the field is free (the point)
					if (checkIfFieldIsFree(newPosX, newPosY)) {
						correctPos = true;
					}

				}
			}
			else {
				correctPos = true;
			}
		}

	} while (!correctPos);

	if (correctPos) {
		newPoint.setX(newPosX);
		newPoint.setY(newPosY);
		return newPoint;
	}
	else {
		//no place around, point stays the same
		return *organismPos;
	}


}

//checking if point is on the map
bool Organism::checkPointCorrenctness(int x, int y) {
	if (y < organismWorld->getHeight() && y >= NULL) {
		if (x < organismWorld->getWidth() && x >= NULL) {
			return true;
		}
	}
	return false;
}

//checking if given field is not occupied by other organism
bool Organism::checkIfFieldIsFree(int x, int y) {

	if (organismWorld->getStatesOfFields()[y][x] != nullptr) {
		return false;
	}
	return true;

}

//starting game function, creating every kind of organism once
void Organism::createEveryOrganism(World* world) {

	//creating animals
	Point newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Wolf(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Sheep(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Fox(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Turtle(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Antelope(world, newPoint));

	//creating plants
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Grass(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new SowThistle(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new Guarana(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new BellaDonna(world, newPoint));
	newPoint = world->randomFreePoint();
	world->addOrganismToWorld(new SosnowskysHogweed(world, newPoint));
}

//creating new organism depending on given specie and position
Organism* Organism::createNewOrganism(OrganismSpecie specie, World* world, Point point) {
	switch (specie) {

	case WOLF: return new Wolf(world, point);
	case SHEEP: return new Sheep(world, point);
	case HUMAN: return new Human(world, point);
	case FOX: return new Fox(world, point);
	case TURTLE: return new Turtle(world, point);
	case ANTELOPE: return new Antelope(world, point);
	case GRASS: return new Grass(world, point);
	case SOW_THISTLE: return new SowThistle(world, point);
	case GUARANA: return new Guarana(world, point);
	case BELLADONNA: return new Guarana(world, point);
	case SOSNOWSKYS_HOGWEED: return new Guarana(world, point);

	default: return nullptr;
	}
}

//drawing organism sign
void Organism::draw() {
	cout << organismSign;
}

//saving info about organism to file
void Organism::saveOrganismToFile(FILE* file) const{

	int x = organismPos->getX(), y = organismPos->getY();
	fwrite(&specie, sizeof(OrganismSpecie), 1, file);
	fwrite(&x, sizeof(int), 1, file);
	fwrite(&y, sizeof(int), 1, file);
	fwrite(&organismSign, sizeof(char), 1, file);
	fwrite(&strength, sizeof(int), 1, file);
	fwrite(&initiative, sizeof(int), 1, file);
	fwrite(&whenBorn, sizeof(int), 1, file);
	fwrite(&killed, sizeof(bool), 1, file);

}

//loading info about organism from file
void Organism::loadOrganismFromFile(FILE* file, World* newWorld) {

	int bornRound = NULL;
	fread(&organismSign, sizeof(char), 1, file);
	fread(&strength, sizeof(int), 1, file);
	fread(&initiative, sizeof(int), 1, file);
	fread(&bornRound, sizeof(int), 1, file);
	setWhenBorn(bornRound);
	fread(&killed, sizeof(bool), 1, file);

}